src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextProvider.ForegroundWorkspaceProjectContext.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices.ProjectSystem;
+
+namespace Microsoft.VisualStudio.ProjectSystem.LanguageServices
+{
+    internal partial class WorkspaceProjectContextProvider
+    {
+        /// <summary>
+        ///     Wraps <see cref="IWorkspaceProjectContext"/> and enforces that it's accessed only on the UI thread.
+        /// </summary>
+        private class ForegroundWorkspaceProjectContext : IWorkspaceProjectContext
+        {
+            private readonly IProjectThreadingService _threadingService;
+            private readonly IWorkspaceProjectContext _underlyingContext;
+
+            public ForegroundWorkspaceProjectContext(IProjectThreadingService threadingService, IWorkspaceProjectContext underlyingContext)
+            {
+                Assumes.NotNull(threadingService);
+                Assumes.NotNull(underlyingContext);
+
+                _threadingService = threadingService;
+                _underlyingContext = underlyingContext;
+            }
+
+            public IWorkspaceProjectContext UnderlyingContext
+            {
+                get
+                {
+                    _threadingService.VerifyOnUIThread();
```

Comment:
Why not just switch to the UI thread and return a task?

Reply:
1) I'm implementing an existing interface that I have no control over.
2) I very deliberately coordinate changes to the context - I don't want random consumers switching to the UI and applying things out of order. Once Jason's work is in, this will be free-threaded and we won't need to enforce this.